theano/scan_module/tests/test_scan_utils.py

```diff
@@ -11,3 +15,146 @@ def test_equal_compuations():
+        outer = tensor.scalar("outer")
+        shared = theano.shared(1, name="shared")
+        constant = tensor.constant(1, name="constant")
+
+        # z will equal 1 so multiplying by it doesn't change any values
+        z = outer * (shared + constant)
+
+        def step(x, a):
+            r = a + x
+            r.tag.replacement = z * (a - x)
+            return r
+
+        s, _ = theano.scan(step, sequences=x,
+                           outputs_info=[numpy.array(0.)])
+        # ensure z is owned by the outer graph so map_variables() will need to
+        # jump through additional hoops to placate FunctionGraph.
+        t = z * s
+        s2, = map_variables(self.replacer, [t])
+        t2 = z * s2
+
+        f = theano.function([x, outer], [t, t2])
+        rval = f(x=numpy.array([1, 2, 3], dtype=numpy.float32), outer=0.5)
+        assert numpy.array_equal(rval, [[1, 3, 6], [-1, -3, -6]])
+
+    def test_scan_with_shared_update(self):
+        x = tensor.vector('x')
+
+        # counts how many times its value is used
+        counter = theano.shared(0, name="shared")
+        counter.update = counter + 1
+
+        def step(x, a):
+            r = a + x
+            # introducing a shared variable with an update into the
+            # inner graph is unsupported and the code must crash rather
+            # than silently produce the wrong result.
+            r.tag.replacement = counter * (a - x)
+            return r
+
+        s, _ = theano.scan(step, sequences=x,
+                           outputs_info=[numpy.array(0.)])
+        try:
+            s2, = map_variables(self.replacer, [s])
+        except NotImplementedError as e:
+            e = sys.exc_info()[1]
+            assert("introduces shared variable" in str(e))
+            return
+
+        # test failed
+        return 0
+
+    def test_scan_with_shared_update3(self):
+        x = tensor.vector('x')
+
+        # counts how many times its value is used
+        counter = theano.shared(0, name="shared")
+        counter.update = counter + 1
+
+        def step(x, a):
+            r = a + x
+            # introducing a shared variable with an update into the
+            # inner graph is unsupported and the code must crash rather
+            # than silently produce the wrong result.
+            r.tag.replacement = counter * (a - x)
+            # the shared variable was already present, but the
+            # replacement changes the number of times it is used,
+            # which would have to change the updates, which is
+            # unsupported.
+            return r + counter
+
+        s, _ = theano.scan(step, sequences=x,
+                           outputs_info=[numpy.array(0.)])
+        try:
+            s2, = map_variables(self.replacer, [s])
+        except NotImplementedError as e:
+            e = sys.exc_info()[1]
+            assert("introduces shared variable" in str(e))
+            return
+
+        # test failed
+        return 0
```

Comment:
You can replace, from the try to the return, by a call to self.assertRaises(NotImplementedError, map_variables, self.replacer, [s]).
For self.assertRaises to be defined, your test class must inherit from unittest.TestCase

This comment also applies to the other test that follows the same pattern.